connector/keystone/keystone.go

```diff
@@ -310,3 +351,129 @@ func (p *conn) getUserGroups(ctx context.Context, userID string, token string) (
 	}
 	return groups, nil
 }
+
+func (p *conn) groupsRequired(groupScope bool) bool {
+	return len(p.Groups) > 0 || groupScope
+}
+
+// If project ID is left empty, all roles will be fetched
+func (p *conn) getUserRolesAsGroups(ctx context.Context, token string, userID string, projectID string) ([]string, error) {
+	roleAssignments, err := p.getRoleAssignments(ctx, token, userID, projectID)
+	if err != nil {
+		return nil, err
+	}
+	roles, err := p.getRoles(ctx, token)
+	if err != nil {
+		return nil, err
+	}
+	roleMap := map[string]role{}
+	for _, role := range roles {
+		roleMap[role.ID] = role
+	}
+	var groups []string
+	for _, roleAssignment := range roleAssignments {
+		role, ok := roleMap[roleAssignment.Role.ID]
+		if !ok {
+			// Ignore role assignments to non-existent roles (shouldn't happen)
+			continue
+		}
+		groups = append(groups, role.Name)
+	}
+	return groups, nil
+}
+
+func (p *conn) getRoleAssignments(ctx context.Context, token string, userID string, projectID string) ([]roleAssignment, error) {
+	// https://docs.openstack.org/api-ref/identity/v3/?expanded=validate-and-show-information-for-token-detail,list-role-assignments-detail#list-role-assignments
+	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/v3/role_assignments?effective&user.id=%s&scope.project.id=%s", p.Host, userID, projectID), nil)
```

Comment:
When a user is admin on one tenant and \_member\_ on another tenant, bouncer ends up granting admin privileges on the workload cluster. Can you confirm if that is case with bouncer and if this will work the same way?

Reply:
Yes, also in this case it will get both the group for admin and _member_